src/ucp/tag/offload.h

```diff
@@ -87,4 +90,55 @@ ucp_tag_offload_try_cancel(ucp_worker_t *worker, ucp_request_t *req, int force)
     }
 }
 
+/* For homogeneous clusters it is expected that tag offload messages will arrive
+ * from the single interface. Otherwise need to add a particular iface to tags
+ * hash, where key is a tag masked with tag_sender mask. */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_unexp(ucp_worker_iface_t *iface, ucp_tag_t tag)
+{
+    ucp_worker_t *worker   = iface->worker;
+    ucp_context_t *context = worker->context;
+    ucp_tag_t tag_key;
+    khiter_t hash_it;
+    int ret;
+
+    ++iface->proxy_recv_count;
+
+    if (ucs_unlikely(!(iface->flags & UCP_WORKER_IFACE_FLAG_ON_OFFLOAD_QUEUE))) {
+
+        if (ucs_unlikely(!context->config.ext.tm_offload)) {
```

Comment:
the internal code block could be placed in a non-inline function to reduce register spill